### ISSUE TITLE
Potential crash on STRtree

### ISSUE THREAD

**Issue by SongHerz** [Toblerity/Shapely#505]
Hi,

I am using shapely STRtree in a project, sometimes it gives wrong result, sometimes it crashes the Python interpreter.

I find that, geoms are not stored in a STRtree instance after initialization. And at query time, pointers to these geoms may not be valid when no other object reference them.

And it looks the issue still exists in the master branch, by reading the code.

## Steps to reproduce the problem.

Here is a small example to reproduce the issue by force gc.

```python
#!/usr/bin/env python

import gc

from shapely.geometry import box
from shapely.strtree import STRtree


tree = STRtree([box(0, 0, 1, 1)])
gc.collect()
result = tree.query(box(0, 0, 1, 1))
for each in result:
    print each
```

The output is:
```
(<nil>,)
```

The expected output is:

```
POLYGON ((1 0, 1 1, 0 1, 0 0, 1 0))
```

By adding `self._geoms = geoms` after the line `geoms = [geom for geom in geoms if not geom.is_empty]` to strtree.py:STRtree:\_\_init\_\_

The issue solved.

## Operating system

Ubuntu 14.04.5 LT

## Shapely version and provenance

Shapely==1.5.17.post1 installed from PyPI using pip

**Comment by snorfalorpagus** [comment]
Thanks for the good report @SongHerz!

This occurs because we're passing pointers to Python objects (geometries) into GEOS without incrementing their reference count. When their reference count drops to 0, which happens immediately in your example as you don't keep a reference to the Polygon you create, the garbage collector frees the object (which will free the underlying GEOSGeometry also).

Here's a slightly more clear example, which causes a segfault on my machine:

```
import gc
from shapely.geometry import box, Point
from shapely.strtree import STRtree

# create a geometry
g1 = Point(0.5, 0.5)
g2 = Point(1.5, 1.5)

# add it to an STRtree
tree = STRtree([g1, g2])

# delete the geometry (and garbage collect)
g1 = None
g2 = None
gc.collect()

# query the tree
result = tree.query(box(0, 0, 1, 1))
for x in result:
    print(x)
```

As you mention, keeping a reference to the list of geometries fixes the problem. This is because there is still a (correctly counted) reference to the objects. The other way to do this would be to manually increment the count on the objects with `ctypes.pythonapi.Py_IncRef(pyo)`, but this would also require manually decrementing when the STRtree object is destroyed with a corresponding `Py_DecRef`. I don't know if there is any benefit to using the more complicated approach?

**Comment by sgillies** [comment]
Interesting. @snorfalorpagus maybe the tree had better keep references to added objects?

**Comment by sgillies** [comment]
I've fixed this in 604eb9f.
